File: python/143. Reorder List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow, fast = head, head.next
        while fast != None and fast.next != None:
            slow, fast = slow.next, fast.next.next
        second = slow.next
        slow.next = None
        prev = None
        while second:
            temp = second.next
            second.next = prev
            prev = second
            second = temp
        head2 = head
        while prev:
            temp = head2.next
            temp2 = prev.next
            head2.next = prev
            prev.next = temp
            head2 = temp
            prev = temp2
        

File: go/143. Reorder List.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reorderList(head *ListNode)  {
    slow, fast := head, head.Next;
    for fast != nil && fast.Next != nil {
        slow, fast = slow.Next, fast.Next.Next;
    }
    second := slow.Next;
    slow.Next = nil;
    var prev *ListNode
    for second != nil {
        temp := second.Next
        second.Next = prev
        prev = second
        second = temp
    }
    head2 := head
    for prev != nil {
        temp := head2.Next;
        temp2 := prev.Next
        head2.Next = prev
        prev.Next = temp
        head2 = temp
        prev = temp2
    }
}

